tmt/steps/provision/__init__.py

```diff
@@ -579,13 +651,38 @@ def _query_is_ostree(self, guest: 'Guest') -> Optional[bool]:
 
         return output.stdout.strip() == 'yes'
 
-    def _query_capabilities(self, guest: 'Guest') -> dict[GuestCapability, bool]:
-        # TODO: there must be a canonical way of getting permitted capabilities.
-        # For now, we're interested in whether we can access kernel message buffer.
-        return {
-            GuestCapability.SYSLOG_ACTION_READ_ALL: True,
-            GuestCapability.SYSLOG_ACTION_READ_CLEAR: True
-            }
+    def _query_capabilities(self, guest: 'Guest') -> Optional[list[GuestCapability]]:
+        """
+        Read /proc/self/status to determine the permitted capabilities.
+
+        The CapPrm line in /proc/self/status contains a hex number representing
+        a bitmask of permitted capabilities. Each bit position corresponds to
+        a capability's index as defined in the kernel's capabilities.h.
+
+        :returns: A list of GuestCapability enum members representing the currently
+                 permitted capabilities, or an empty list if capabilities cannot be determined.
+        """
+        try:
+            output = self._execute(guest, Command('grep', 'CapPrm', '/proc/self/status'))
+        except tmt.utils.RunError:
+            return []
+
+        if not output or not output.stdout:
+            return []
+
+        # Extract the permitted bitmask from the CapPrm line
+        bitmask_hex = output.stdout.split()[1]
+        # TODO If hex = 0000000000000000, we could try again to read /proc/self/status with
+        # sudo to get capabilities available for superuser
+        bitmask = int(bitmask_hex, 16)
+
+        # Convert the bitmask to a list of permitted capabilities
+        capabilities: list[GuestCapability] = []
+        for cap in GuestCapability:
+            if bitmask & (1 << cap.value):
+                capabilities.append(cap)
```

Comment:
List comprehension to the rescue! Or at least to a slightly shorter & more elegant oneliner without an accumulator variable:

```python
return [cap for cap in GuestCapability if bitmask & (1 << cap.value)]
```

Reply:
Weird that wasn't caught by ruff.

Reply:
Addressed in 62ad828b667fc6680072f375e6bb30edff9725a9